src/core/native/adapter.ts

```diff
@@ -10,5 +11,7 @@ export interface Adapter {
   visitRequestFailedWithStatusCode(visit: Visit, statusCode: number): void
   visitRequestFinished(visit: Visit): void
   visitRendered(visit: Visit): void
+  formSubmissionStarted(formSubmission: FormSubmission): void
+  formSubmissionFinished(formSubmission: FormSubmission): void
```

Comment:
@sstephenson @javan what is necessary for parity across Web/iOS/Android when changing the adapter interface?

If it requires a lot of effort, maybe it would be worthwhile to unify the `visitRequestStarted` and `visitRequestFinished` along with the new `formSubmissionStarted` and `formSubmissionFinished` methods and replace them with something more general like `fetchRequestStarted` and `fetchRequestEnded`?

Reply:
We had previously proposed that we use `[data-turbo-busy]` on all elements as an indicator that a request is in-flight (https://github.com/hotwired/turbo/pull/199). If we were able to consistently set that attribute, we could manage the visibility and animation of a `<progress>` or `<div role="progress">` element through CSS or a `MutationObserver` instance.

That might be a more simplified implementation.